### ISSUE TITLE
return ABC particle distances

### ISSUE THREAD

**Issue by robinmarin** [stewid/SimInf#40]
When determining a good tolerance vector, getting feedback from the resulting 
abc-SMC particles would be very much appreciated. Instead of "hacking" this into 
the acceptance function, it would be nice if the computed distances of accepted particles
could be included in the output object. Furthermore, adaptively selecting the tolerance from 
the quantiles of accepted samples could then also, be implemented within the solver. 

TLDR;
1. distances of accepted ABC particles in the output object like the weights.
2. Implement adaptivity in the tolerance from the distances of accepted particles.

**Comment by trosendal** [comment]
Interesting suggestion! I can say that work that @KSTuominen and I are doing would also benefit from such a feature. We usually supply a vector of tolerances and let the algorithm run until it gets 'stuck' on a generation. Then we stop the ABC and back off the tolerance a little and restart it from the previous generation. An adaptive approach would be very practical and easier to describe in a methods section of course :)

**Comment by stewid** [comment]
@robineriksson Thanks for the suggestion. I've added a new branch `ABC-particle-distance` where I've added a slot `distance` to the `SimInf_abc` class. Any thoughts on how to inform the algorithm to use adaptivity for the tolerance? Maybe one could return a different result from the accept function; currently `abc_accept(accept, epsilon)` is the only valid result, but maybe one could also have `abc_accept_adaptive(distance)`?

**Comment by robinmarin** [comment]
hm, well `abc_accept(accept , distance)` for both cases: (1) only store distances for review, and (2) potential adaptivity. But then flag adaptivity already in the starting call, like `abc(..., adaptive=TRUE)`. Because the adaptivity should probably be considered after the particles for the generation has been accepted, and it would help with potential confusion when user writes `acceptFun`.

**Comment by stewid** [comment]
Yes, I agree, adding `adaptive` as an argument to the call `abc(..., adaptive = TRUE)` is better. Another improvement in writing the `acceptFun` is to change it to only return the `distance` by adding the tolerance vector `epsilon` as an argument to `abc(..., epsilon, adaptive)`. Then the solver can use that vector to determine if a particle should be accepted or not.

**Comment by robinmarin** [comment]
Yes, good catch. That would be a nice improvement.

**Comment by stewid** [comment]
I've merged the branch `ABC-particle-distance` into `main`. The particle distances are now collected in the `distance` slot. I've also added the argument tolerance `abc(..., tolerance)`. Additionally, the distance should now only return the distance for each particle. The next step is to handle potential adaptivity. @robineriksson any thoughts on a suitable algorithm? I found a paper `Adaptive Approximate Bayesian Computation Tolerance Selection` (https://doi.org/10.1214/20-BA1211) that I'm planning to read.

**Comment by robinmarin** [comment]
Nice! 

Yes, `Adaptive Approximate Bayesian Computation Tolerance Selection` is a good read. It also cites older approaches to adaptive tolerances. The method they propose is also promising and worth taking a look at. I read it as such: they find the new tolerances from the quantiles as earlier approaches **but** they also let the quantile in question level be adaptively chosen.

**Comment by stewid** [comment]
I've now implemented adaptive tolerance selection using the algorithm of Simola and others (2021). To use it, one must specify the number of particles (`ninit`) to run in the first generation, see example below. My plan is to submit a new version to CRAN this week.

``` r
library(SimInf)

## Let us consider an SIR model in a closed population with N = 100
## individuals of whom one is initially infectious and the rest are
## susceptible. First, generate one realisation (with a specified
## seed) from the model with known parameters \code{beta = 0.16} and
## \code{gamma = 0.077}. Then, use \code{abc} to infer the (known)
## parameters from the simulated data.
model <- SIR(u0 = data.frame(S = 99, I = 1, R = 0),
             tspan = 1:100,
             beta = 0.16,
             gamma = 0.077)

## Run the SIR model and plot the number of infectious.
set.seed(22)
infectious <- trajectory(run(model), "I")$I

## The distance function to accept or reject a proposal. Each node
## in the simulated trajectory (contained in the 'result' object)
## represents one proposal.
distance <- function(result, ...) {
    ## Extract the time-series of infectious in each node as a
    ## data.frame.
    sim <- trajectory(result, "I")

    ## Split the 'sim' data.frame by node and calculate the sum of the
    ## squared distance at each time-point for each node.
    dist <- tapply(sim$I, sim$node, function(sim_infectious) {
        sum((infectious - sim_infectious)^2)
    })

    ## Return the distance for each node. Each proposal will be
    ## accepted or rejected depending on if the distance is less than
    ## the tolerance for the current generation.
    dist
}

## Fit the model parameters using ABC-SMC and adaptive tolerance
## selection. The priors for the parameters are specified using a
## formula notation. Here we use a uniform distribtion for each
## parameter with lower bound = 0 and upper bound = 1.
fit <- abc(model = model,
           priors = c(beta ~ uniform(0, 1), gamma ~ uniform(0, 1)),
           npart = 1000,
           ninit = 10000,
           distance = distance)

## Print a brief summary.
fit
#> Number of particles per generation: 1000
#> Number of generations: 7
#> 
#> Generation 7:
#> -------------
#>  Accrate: 5.40e-02
#>  ESS: 6.98e+02
#> 
#>          Min. 1st Qu. Median   Mean 3rd Qu.   Max.
#>  beta  0.0238  0.1331 0.1672 0.1689  0.2022 0.3723
#>  gamma 0.0208  0.0698 0.0839 0.0852  0.1001 0.1645

## Display the ABC posterior distribution.
plot(fit)
```

![](https://i.imgur.com/eFLHYGf.png)

<sup>Created on 2022-04-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by stewid** [comment]
A new version (v9.0.0) with the improved functionality is now on CRAN.
